src/System.Private.CoreLib/shared/System/Text/Utf8Utility.Validation.cs

```diff
@@ -0,0 +1,659 @@
+                        asciiAdjustment = (nint)(sbyte)thisDWord >> 7; // smear most significant bit of least significant byte across entire value
+                    }
+
+                    // asciiAdjustment = 0 if fourth byte is ASCII; -1 otherwise
+
+                    // The two lines below *MUST NOT* be reordered. It's valid to add 4 before backing up since we
+                    // already checked previously that the input buffer contains at least a DWORD's worth of data,
+                    // so we're not going to run past the end of the buffer where the GC can no longer track the
+                    // reference. However, we can't back up before adding 4, since we might back up to before the
+                    // start of the buffer, and the GC isn't guaranteed to be able to track this.
+
+                    inputBuffer = ref Unsafe.Add(ref inputBuffer, 4); // optimisitcally, consumed a 3-byte UTF-8 sequence plus an extra ASCII byte
+                    inputBuffer = ref Unsafe.Add(ref inputBuffer, (IntPtr)asciiAdjustment); // back up if we didn't actually consume an ASCII byte
+
+                    tempUtf16CodeUnitCountAdjustment -= 2; // 3 (or 4) UTF-8 bytes -> 1 (or 2) UTF-16 code unit (and 1 [or 2] scalar)
+
+                SuccessfullyProcessedThreeByteSequence:
+
+                    // Optimization: A three-byte character could indicate CJK text, which makes it likely
+                    // that the character following this one is also CJK. We'll try to process several
+                    // three-byte sequences at a time.
+
+                    // The check below is really "can we read 9 bytes from the input buffer?" since 'finalPos...' is already offset
+                    if (IntPtr.Size >= 8 && BitConverter.IsLittleEndian && (nint)(void*)Unsafe.ByteOffset(ref inputBuffer, ref finalPosWhereCanReadDWordFromInputBuffer) >= 5)
+                    {
+                        ulong thisQWord = Unsafe.ReadUnaligned<ulong>(ref inputBuffer);
+
+                        // Is this three 3-byte sequences in a row?
+                        // thisQWord = [ 10yyyyyy 1110zzzz | 10xxxxxx 10yyyyyy 1110zzzz | 10xxxxxx 10yyyyyy 1110zzzz ] [ 10xxxxxx ]
+                        //               ---- CHAR 3  ----   --------- CHAR 2 ---------   --------- CHAR 1 ---------     -CHAR 3-
+                        if ((thisQWord & 0xC0F0C0C0F0C0C0F0ul) == 0x80E08080E08080E0ul && UnicodeUtility.IsUtf8ContinuationByte(in Unsafe.Add(ref inputBuffer, 8)))
+                        {
+                            // Saw a proper bitmask for three incoming 3-byte sequences, perform the
+                            // overlong and surrogate sequence checking now.
+
+                            // Check the first character.
+                            // If the first character is overlong or a surrogate, fail immediately.
+
+                            if ((((uint)thisQWord & 0x200Fu) == 0) || ((((uint)thisQWord - 0x200Du) & 0x200Fu) == 0))
```

Comment:
This check seems to be repeated quite often. Extract it out to a helper method?

Reply:
Can't, unfortunately. Today's JIT can't fold complex branches when they exist inside inlined methods. Related to https://github.com/dotnet/coreclr/issues/914.